flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java

```diff
@@ -65,7 +67,9 @@
  *
  * @param <IN> The type of the input elements.
  */
-class KafkaWriter<IN> implements SinkWriter<IN, KafkaCommittable, KafkaWriterState> {
+class KafkaWriter<IN>
+        implements SinkWriter<IN, KafkaCommittable, KafkaWriterState>,
+                MetadataPublisher<RecordMetadata> {
```

Comment:
Actually, my initial idea was to implement the `MetadataPublisher` with the Sink and only pass the list of subscribers to the writer.

I am not sure about the final design of the table store but my assumption was that there are only new subscribers during the writer creation and not during runtime. So exposing adding new subscribers through the writer is unnecessary.
Does that make sense?

Reply:
The big question is whether the `subscribe` call occurs before or after serialization